### PR TITLE
inline UI helper text and tracking (source mode)

### LINKED ISSUE #1026: Source UI: clarify meaning of buffer size
Buffer size is mysterious:
screen shot 2014-11-17 at 1 53 37 pm

Inline copy? Something like:

> Increase the buffer size to increase the flexibilty of label, marker, and shield symbolizer placement at the cost of speed and efficiency.

A UI component that visualizes buffer size?
image3540

### LINKED ISSUE #1027: Source UI: clarify meaning of minzoom/maxzoom
These innocuous looking sliders are powerful / dangerous:
screen shot 2014-11-17 at 2 25 28 pm
- Add explanatory copy.
- Visualize or display the size (in tiles?) of export for user as user changes export area.

### PR DESCRIPTION
**Pull request by amyleew**
Adding inline helper text to Studio UI in source mode - specifically for minzoom, maxzoom and buffer definition and application.

image

image

Closes #1027
Closes #1026

### CODE REVIEW

**Comment by amyleew** [comment]
Other options for more copy and scalability.

### Click reveal

expand_click

### Open docs tab

image

04_docs-tab

**Comment by samanpwbb** [comment]
I definitely agree that inline text is the approach we should take at least to start. Looks great to me. Make sure you update all the source templates!

**Comment by amyleew** [comment]
Updated copy - take a look before I roll out to other formats @ajashton @samanpwbb @yhahn 

image

image

**Comment by amyleew** [comment]
Added [on source SAVE tracking](https://github.com/mapbox/mapbox-studio/blob/1df75ff6ac49699440afefd9f7f8ddf16ce2e33c/app/source.js#L539-L551) to:
- layer count
- buffer size
- max and minzoom settings
- file size

Changed both min and maxzoom to max out at 16 vs. 22:

image

/cc @yhahn @ajashton @samanpwbb

File: app/source.js [hunk 1]
```diff
@@ -534,6 +535,21 @@ window.Source = function(templates, cwd, tm, source, revlayers, examples, isMapb
...
+		analytics.track('buffer layer', { buffer: attr.layername });
+		analytics.track('buffer size setting', { bufferlayer: attr.layer.buffer-size });
+
+		// Track max and min zooms
+        analytics.track('maxzoom setting', { maxzoom: attr.maxzoom });
+        analytics.track('minzoom setting', { minzoom: attr.minzoom });
+
+		// Track file size on save
+		analytics.track('file size', { filesize: filesize });
+
```
Comment (yhahn) [inline app/source.js]: want to consistify the indenting here? Studio is using 4 spaces in many places for tabs.

File: app/source.js [hunk 2]
```diff
@@ -574,13 +590,13 @@ window.Source = function(templates, cwd, tm, source, revlayers, examples, isMapb
...
         map.options.minZoom = this.model.get('minzoom');
-        map.options.maxZoom = 22;
+        map.options.maxZoom = 16;
         // Refresh map layer.
         if (tiles) map.removeLayer(tiles);
         tiles = L.mapbox.tileLayer({
             tiles: ['/source/{z}/{x}/{y}.png?id=' + this.model.id + '&' + mtime],
             minzoom: this.model.get('minzoom'),
-            maxzoom: 22
+            maxzoom: 16
```
Comment (yhahn) [inline app/source.js]: These maxzoom values controls the max zoom that the source editor allows you to preview your source at. I think it probably makes sense to leave this at z22 (ie. so you can see what your source data looks like overzoomed).

To limit the maxzoom that the sliders in the settings pane allow I think you'll want to adjust the max properties here:

https://github.com/mapbox/mapbox-studio/blob/mb-pages/templates/source.html#L91-L104

File: app/source.js [hunk 3]
```diff
@@ -534,6 +535,21 @@ window.Source = function(templates, cwd, tm, source, revlayers, examples, isMapb
...
+        analytics.track('buffer layer', { buffer: attr.layername });
+        analytics.track('buffer size setting', { bufferlayer: attr.layer.buffer-size });
+
+        // Track max and min zooms
+        analytics.track('maxzoom setting', { maxzoom: attr.maxzoom });
+        analytics.track('minzoom setting', { minzoom: attr.minzoom });
+
+        // Track file size on save
+        analytics.track('file size', { filesize: filesize });
+
```
Comment (yhahn) [inline app/source.js]: Per chat, 
- These will need to be above the `return` on line 537 to get called.
- `attr.layer.buffer-size` won't work in js -- you'll want to access it with the brackets + quotes (http://stackoverflow.com/a/7122629)

**Comment by amyleew** [comment]
@yhahn Finally got it tracking minzoom, maxzoom and buffer sizes per layers on zoom. thanks @samanpwbb's tutelage on console.log and objects 

image

image

**Comment by yhahn** [comment]
:ship:

**Comment by yazjisuhail** [comment]
@amyleew why did you guys lock the max zoom to 16 in the latest version?

**Comment by yazjisuhail** [comment]
@amyleew modifying the max zoom in the source project's yml won't load the data beyond 16. does it need any kind of refresh/restart?

**Comment by amyleew** [comment]
@yazjisuhail Close Studio and use a text editor (like Sublime) to modify the `maxzoom` in the `data.yml` in your project's .tm2source folder. Restart Studio and your file's `maxzoom` will be changed.

**Pull request merged**